src/mixins/keyboard-shortcut-mixin.ts

```diff
@@ -1,26 +1,65 @@
 import { LitElement } from "lit-element";
 import { Constructor } from "../types";
 
+type RegisteredShortcuts = {
+  [key in HAKeyboardShortcut]?: () => void;
+};
+
+export enum HAKeyboardShortcut {
+  CTRL_S = "CtrlS",
+  CTRL_P = "CtrlP",
+  CTRL_SHIFT_P = "CtrlShiftP",
+}
+
+const isMac = /(Mac|iPhone|iPod|iPad)/i.test(navigator.platform);
+
 export const KeyboardShortcutMixin = <T extends Constructor<LitElement>>(
   superClass: T
 ) =>
   class extends superClass {
-    private _keydownEvent = (event: KeyboardEvent) => {
-      if ((event.ctrlKey || event.metaKey) && event.key === "s") {
-        event.preventDefault();
-        this.handleKeyboardSave();
-      }
-    };
+    private _registeredShortcuts: RegisteredShortcuts = {};
 
-    public connectedCallback() {
-      super.connectedCallback();
-      this.addEventListener("keydown", this._keydownEvent);
+    protected registerShortcut(
+      shortcut: HAKeyboardShortcut,
+      callback: () => void,
+      element: HTMLElement | Document = this
+    ) {
+      element.addEventListener("keydown", this._keydownEvent);
+      this._registeredShortcuts[shortcut] = callback;
     }
 
-    public disconnectedCallback() {
-      this.removeEventListener("keydown", this._keydownEvent);
-      super.disconnectedCallback();
+    private _executeShortcut(
+      event: KeyboardEvent,
+      shortcut: HAKeyboardShortcut
+    ) {
+      if (Object.keys(this._registeredShortcuts).includes(shortcut)) {
+        event.preventDefault();
+        const shortcutAction = this._registeredShortcuts[shortcut];
+        if (shortcutAction) {
+          shortcutAction();
+        }
+      }
     }
 
-    protected handleKeyboardSave() {}
+    private _keydownEvent = (event: Event) => {
+      const _event = event as KeyboardEvent;
```

Comment:
Why not?
```suggestion
    private _keydownEvent = (event: KeyboardEvent) => {
```

Reply:
I uh... er... um... well you see...

Reply:
Done

Reply:
That is strange. If using even a remotely recent version of typescript (i.e. from some point in 2018 onwards) and the first parameter is "keydown"/"keyup"/"keypress", it is supposed to enable an overload with a callback that takes `KeyboardEvent`. You can find the [mapping here](https://github.com/microsoft/TypeScript/blob/7a484695ca3452b29b268d037ce1d7dda9d69761/lib/lib.dom.d.ts#L5717). Warning: Positively enormous file (20k lines long), which could potentially crash a browser.

Wonder if there is some form of overload resolution bug being hit here. Oh well, that workaround is reasonable enough.

I've seen [some indication](https://github.com/jamiebuilds/tinykeys/pull/12) that at least one browser or js framework at one point or other would pass events that were not derived from KeyboardEvent (and not fully duck type compatible) in some edge cases, but the typescript type definitions definitely do not that that into account. I have no positive knowledge of this occurring, so this is just hearsay.